compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -6664,8 +6664,11 @@ public void visit(BLangIndexBasedAccess indexAccessExpr) {
             targetVarRef = new BLangArrayAccessExpr(indexAccessExpr.pos, indexAccessExpr.expr,
                                                     indexAccessExpr.indexExpr);
         } else if (types.isAssignable(varRefType, symTable.xmlType)) {
-            targetVarRef = new BLangXMLAccessExpr(indexAccessExpr.pos, indexAccessExpr.expr,
-                    indexAccessExpr.indexExpr);
+            BLangExpression indexAccessExprExpr = indexAccessExpr.expr;
+            if (Types.getImpliedType(indexAccessExprExpr.getBType()).tag == TypeTags.UNION) {
+                indexAccessExprExpr = createTypeCastExpr(indexAccessExprExpr, symTable.xmlType);
+            }
```

Comment:
Why do we need to create a type cast here only for XML? For other types, if the expression type is a union, we do not create any type cast. Is there any reason to special-case this scenario?

Reply:
We do not support invocation of langlib functions on xml union types. https://github.com/ballerina-platform/ballerina-lang/issues/28697. Thus when we get a union type in order to support the indexed based access we need to type cast it to `xml`.

Reply:
My suggestion is that we do not cast here. Instead, in the `MethodGen`, if the type is a union and a subtype of XML, we can update the logic accordingly.

My concern with creating the cast here is that it is not included in the generated bytecode. Isn't that correct?

Reply:
Had an offline chat and decided to keep this as it and remove the cast once this is addressed. https://github.com/ballerina-platform/ballerina-lang/issues/28697

Reply:
Shall we add a comment (to-do) on the code, by mentioning the issue? If not, we can miss this